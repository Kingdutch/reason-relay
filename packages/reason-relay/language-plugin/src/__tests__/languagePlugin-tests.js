/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @emails oncall+relay
 */

"use strict";

const { buildSchema } = require("graphql");
const fs = require("fs");
const path = require("path");
const RelayReasonGenerator = require("../RelayReasonGenerator");
const { printCode } = require("../generator/Printer.gen");

const GraphQLCompilerContext = require("relay-compiler/lib/core/GraphQLCompilerContext");
const RelayFlowGenerator = require("relay-compiler/lib/language/javascript/RelayFlowGenerator");
const RelayIRTransforms = require("relay-compiler/lib/core/RelayIRTransforms");

const { transformASTSchema } = require("relay-compiler/lib/core/ASTConvert");
const { parseGraphQLText } = require("relay-test-utils-internal");

const testSchema = buildSchema(
  fs.readFileSync(
    path.resolve(path.join(__dirname, "testSchema.graphql")),
    "utf8"
  )
);

import type { TypeGeneratorOptions } from "relay-compiler/lib/RelayLanguagePluginInterface";

function generate(text, options?: TypeGeneratorOptions, extraDefs = "") {
  const schema = transformASTSchema(testSchema, [
    ...RelayIRTransforms.schemaExtensions,
    extraDefs
  ]);
  const { definitions } = parseGraphQLText(schema, text);
  return new GraphQLCompilerContext(testSchema, schema)
    .addAll(definitions)
    .applyTransforms(RelayFlowGenerator.transforms)
    .documents()
    .map(
      doc =>
        `// ${doc.name}.graphql\n${printCode(
          RelayReasonGenerator.generate(doc, {
            customScalars: {},
            optionalInputFields: [],
            existingFragmentNames: new Set([]),
            ...options
          })
        )}`
    )
    .join("\n\n");
}

describe("Language plugin tests", () => {
  describe("Query", () => {
    it("prints the correct operationType type", () => {
      expect(
        generate(
          `query appQuery($userId: ID!) {
            user(id: $userId) {
              id
              firstName
            }
          }`
        ).includes("type operationType = ReasonRelay.queryNode;")
      ).toBe(true);
    });

    it("prints simple responses and variables", () => {
      let generated = generate(
        `query appQuery($userId: ID!) {
            user(id: $userId) {
              id
              firstName
            }
          }`
      );

      expect(generated).toMatchSnapshot();
    });

    it("prints variables as unit if not variables are supplied", () => {
      let generated = generate(
        `query appQuery {
            me {
              id
              firstName
            }
          }`
      );

      expect(generated.includes("type variables = unit;")).toBe(true);
    });

    it("prints nested objects inlined in types", () => {
      let generated = generate(
        `query appQuery($location: LocationBounds!) {
            userByLocation(location: $location) {
              id
              firstName
            }
          }`
      );

      expect(generated).toMatchSnapshot();
    });
  });

  describe("Mutation", () => {
    it("prints the correct operationType type", () => {
      expect(
        generate(
          `mutation SetUserLocationMutation($input: SetUserLocationInput!) {
            setUserLocation(input: $input) {
              changedUser {
                id
                firstName
              }
            }
          }`
        ).includes("type operationType = ReasonRelay.mutationNode;")
      ).toBe(true);
    });
  });

  describe("Subscription", () => {
    it("prints the correct operationType type", () => {
      expect(
        generate(
          `subscription SomeSubscription($input: UserChangedInput!) {
            userChanged(input: $input) {
              user {
                id
                firstName
              }
            }
          }`
        ).includes("type operationType = ReasonRelay.subscriptionNode;")
      ).toBe(true);
    });
  });

  describe("Fragment", () => {
    it("prints the correct operationType type", () => {
      expect(
        generate(
          `fragment SomeComponent_user on User {
            id
            firstName
          }`
        ).includes("type operationType = ReasonRelay.fragmentNode;")
      ).toBe(true);
    });
  });

  describe("Enums", () => {
    it("references any enums by global, generated schema assets file", () => {
      let generated = generate(
        `query appQuery {
            me {
              role
            }
          }`
      );

      expect(
        generated.includes(`"role": SchemaAssets.Enum_UserRole.wrapped`)
      ).toBe(true);
    });
  });

  describe("Unions", () => {
    it("generates code to unwrap unions", () => {
      let generated = generate(
        `query appQuery {
            participantById(id: "123") {
              __typename
              ... on User {
                id
                firstName
                lastName
              }

              ... on Observer {
                id
                name
              }
            }
          }`
      );

      expect(generated).toMatchSnapshot();
    });

    it("generates opaque wrapped union types referenced by path in types", () => {
      let generated = generate(
        `query appQuery {
            participantById(id: "123") {
              __typename
              ... on User {
                id
                firstName
                lastName
              }

              ... on Observer {
                id
                name
              }
            }
          }`
      );

      expect(
        generated.includes("type union_response_participantById_wrapped;")
      ).toBe(true);
    });
  });
});
