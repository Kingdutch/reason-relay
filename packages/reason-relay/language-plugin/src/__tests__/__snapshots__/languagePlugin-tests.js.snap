// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Language plugin tests Fragment prints the correct fragment ref extractor helpers and base types 1`] = `
"// SomeComponent_user.graphql
type t;
type fragmentRef;
type fragmentRefSelector('a) =
  {.. \\"__$fragment_ref__SomeComponent_user\\": t} as 'a;
external getFragmentRef: fragmentRefSelector('a) => fragmentRef = \\"%identity\\";

type fragment = {. \\"id\\": string};
type operationType = ReasonRelay.fragmentNode;

module Unions = {};
"
`;

exports[`Language plugin tests Mutation prints the correct basic structure for mutations 1`] = `
"// SetUserLocationMutation.graphql
type response = {
  .
  \\"setUserLocation\\": {
    .
    \\"changedUser\\":
      Js.Nullable.t({
        .
        \\"firstName\\": string,
        \\"id\\": string,
      }),
  },
};
type variables = {
  .
  \\"input\\": {
    .
    \\"location\\":
      option({
        .
        \\"lng\\": float,
        \\"lat\\": float,
      }),
    \\"userId\\": string,
    \\"clientMutationId\\": option(string),
  },
};
type operationType = ReasonRelay.mutationNode;

module Unions = {};
"
`;

exports[`Language plugin tests Query prints nested objects inlined in types 1`] = `
"// appQuery.graphql
type response = {
  .
  \\"userByLocation\\":
    Js.Nullable.t({
      .
      \\"firstName\\": string,
      \\"id\\": string,
    }),
};
type variables = {
  .
  \\"location\\": {
    .
    \\"lng\\": float,
    \\"lat\\": float,
  },
};
type operationType = ReasonRelay.queryNode;

module Unions = {};
"
`;

exports[`Language plugin tests Query prints simple responses and variables 1`] = `
"// appQuery.graphql
type response = {
  .
  \\"user\\":
    Js.Nullable.t({
      .
      \\"firstName\\": string,
      \\"id\\": string,
    }),
};
type variables = {. \\"userId\\": string};
type operationType = ReasonRelay.queryNode;

module Unions = {};
"
`;

exports[`Language plugin tests Subscription prints the correct basic structure for subscriptions 1`] = `
"// SomeSubscription.graphql
type response = {
  .
  \\"userChanged\\": {
    .
    \\"user\\":
      Js.Nullable.t({
        .
        \\"firstName\\": string,
        \\"id\\": string,
      }),
  },
};
type variables = {
  .
  \\"input\\": {
    .
    \\"userId\\": string,
    \\"clientMutationId\\": option(string),
  },
};
type operationType = ReasonRelay.subscriptionNode;

module Unions = {};
"
`;

exports[`Language plugin tests Unions generates code to unwrap unions 1`] = `
"// appQuery.graphql
type union_response_participantById_wrapped;

type response = {
  .
  \\"participantById\\": Js.Nullable.t(union_response_participantById_wrapped),
};
type variables = unit;
type operationType = ReasonRelay.queryNode;

module Unions = {
  module Union_response_participantById: {
    type type_User = {
      .
      \\"id\\": string,
      \\"firstName\\": string,
      \\"lastName\\": string,
    };
    type type_Observer = {
      .
      \\"id\\": string,
      \\"name\\": string,
    };
    type t = [
      | \`User(type_User)
      | \`Observer(type_Observer)
      | \`UnmappedUnionMember
    ];
    let unwrap: union_response_participantById_wrapped => t;
  } = {
    external __unwrap_union:
      union_response_participantById_wrapped => {. \\"__typename\\": string} =
      \\"%identity\\";
    type type_User = {
      .
      \\"id\\": string,
      \\"firstName\\": string,
      \\"lastName\\": string,
    };
    type type_Observer = {
      .
      \\"id\\": string,
      \\"name\\": string,
    };
    type t = [
      | \`User(type_User)
      | \`Observer(type_Observer)
      | \`UnmappedUnionMember
    ];
    external __unwrap_User: union_response_participantById_wrapped => type_User =
      \\"%identity\\";
    external __unwrap_Observer:
      union_response_participantById_wrapped => type_Observer =
      \\"%identity\\";
    let unwrap = wrapped => {
      let unwrappedUnion = wrapped |> __unwrap_union;
      switch (unwrappedUnion##__typename) {
      | \\"User\\" => \`User(wrapped |> __unwrap_User)
      | \\"Observer\\" => \`Observer(wrapped |> __unwrap_Observer)
      | _ => \`UnmappedUnionMember
      };
    };
  };
};
"
`;
